src/arc_utilities/ros_helpers.py

```diff
@@ -29,10 +29,67 @@ def callback(self, msg):
         with self.lock:
             self.data = msg
 
-    def get(self):
+    def get(self, block_on_none=False):
         """
         Returns the latest msg from the subscribed topic
+
+        Parameters:
+        block_on_none (bool): block return until data has been received
         """
+        wait_for(lambda: not (block_on_none and self.data is None))
+            
         with self.lock:
             return self.data
+
     
+def wait_for(func):
+    """
+    Waits for function evaluation to be true. Exits cleanly from ros.
+
+    Introduces sleep delay, not recommended for time critical operations
+    """
+    
+    while not func() and not rospy.is_shutdown():
+        time.sleep(0.01)
+
+def joy_to_xbox(joy):
+    """
+    Transforms a joystick sensor_msg to a XBox controller for easier code readability
+    
+    Parameters:
+    joy (sensor_msgs/Joy): xbox msg
+
+    Returns:
+    xbox struct where fields are the button names
+    """
+    class Xbox_msg():
+        pass
+    x = Xbox_msg()
+    x.A, x.B, x.X, x.Y, x.LB, x.RB, \
+        x.back, x.start, x.power,\
+        x.stick_button_left, x.stick_button_right, \
+        x.DL, x.DR, x.DU, x.DD = joy.buttons
+    x.LH, x.LV, x.LT, x.RH, x.RV, x.RT, x.DH, x.DV = joy.axes
+    return x
+        
+
+
+class Xbox():
+    def __init__(self):
+        self.xbox_listener = Listener("/joy", Joy)
+
+    def get_buttons_state(self):
+        return joy_to_xbox(self.xbox_listener.get(True))
+
+    def wait_for_button(self, button, message=True):
+        """
+        Waits for button press on xbox.
+
+        Parameters:
+        button (str):   Name of xbox button. "A", "B", "X", ...
+        message (bool): log a message informing the user?
+        """
+        if message:
+            rospy.loginfo("Waiting for xbox button: " + button)
+            
+        wait_for(lambda: getattr(self.get_buttons_state(), button))
```

Comment:
This line of code is not obvious to me, but maybe that's because of a lack of knowledge of python + the way the joy messages are formatted? Are the various buttons boolean types?

Reply:
http://docs.ros.org/api/sensor_msgs/html/msg/Joy.html

buttons are int32 that are 0 or 1 for the xbox controller.

Reply:
Can we add a brief comment here to make this a bit easier to parse if someone has to change something for some reason?
```Python
# Button state is 0 if not pressed, and 1 if pressed, so wait for the state to be 1 before returning
```
or something like that.

Reply:
Even better, I'll make that line a named function. It seems useful outside of the "wait_for_button" loop.